Automattic/wp-calypso
# Gutenboarding: Domain Picker Modal

## Changes proposed in this Pull Request

This PR introduces **Domain Picker Modal** which is enabled via the `gutenboarding/domain-picker-modal` feature flag.

There are some architectural decisions being made here so feel free to review & discuss them in this PR.

### The component tree looks like this:

```
DomainPickerButton
   |_DomainPickerPopover
     |_DomainPicker
   |_DomainPickerModal
     |_DomainPicker
```

### Here are the list of changes:

* **DomainPickerModal component.**
   This new component uses the `react-modal` component and reuses the `DomainPicker` component. It reuses `<html>` element's scrollbar to scroll down modal's content while keeping the gutenboarding header visible. https://github.com/Automattic/wp-calypso/pull/41212/files#diff-3b763db88828368a2cd772bdcb184e7bR1-R24
* **DomainPickerPopover component.**  
   This is refactored from the existing code in `DomainPickerButton` component.
* **New props showDomainConnectButton & showDomainCategories introduced in DomainPicker component.** 
    The parent component DomainPickerPopover and DomainPickerModal decides what to show and what to hide. https://github.com/Automattic/wp-calypso/pull/41212/files#diff-9fb3f50c91d02f132bfab858e87cb792R74-R93
* **Close Button & More Options Button is now outside of DomainPicker component**
   These 2 buttons are specific to DomainPickerPopover, therefore it make sense that DomainPickerPopover component should play the role of decorating the DomainPicker component with these 2 additional buttons. https://github.com/Automattic/wp-calypso/pull/41212/files#diff-3acfaab9099bccb32e73498b4be055a4R55-R72 This also fixes the accessibility issue with close button needing a `tabIndex={ -1 }` to fix incorrect autofocusing. p1586489720342200-slack-gutenboarding

### Other changes:

* **onClose event removed from CloseButton component**
  onClose is basically the same as onClick. It was removed to prevent typing issues as `onClose` did not accept undefined values. This component is only used by DomainPicker component so far.

## Testing instructions

Note: You need to test this locally. Can't access the modal via the live test page as this is hiding behind a feature flag.

1. Go to `/new` and enter your site name.
2. Click on domain picker button.
3. Domain picker popover should appear.
    - Due to refactoring, please test all functionalities of domain picker inside a popover, it should behave like before.
    - You should now be able to cycle-tab all the form elements in the domain picker popover, including the close button.
4. Click on the **More Options** button
5. Domain picker modal should appear.
    - Domain picker button should disappear.
    - Resize the browser viewport height until a scrollbar appears.
    - Try scrolling the modal content.
6. Try opening the domain picker modal on every other gutenboarding pages.

## Screenshot

![image](https://user-images.githubusercontent.com/1287077/79568731-f803fb80-80e8-11ea-8f9e-cf5f075ad66d.png)

Fixes #41081
